src/otaclient/ota_core.py

```diff
@@ -405,11 +402,62 @@ def _process_persistents(self, ota_metadata: ota_metadata_parser.OTAMetadata):
                 continue
 
             try:
-                _handler.preserve_persist_entry(_per_fpath)
+                _handler.preserve_persist_entry(persiste_entry)
             except Exception as e:
-                _err_msg = f"failed to preserve {_per_fpath}: {e!r}, skip"
+                _err_msg = f"failed to preserve {persiste_entry}: {e!r}, skip"
                 logger.warning(_err_msg)
 
+    def _download_and_parse_metadata(self) -> None:
+        self._download_watchdog_ctx["previous_active_timestamp"] = int(time.time())
+        _mapper = ThreadPoolExecutorWithRetry(
+            max_concurrent=cfg.MAX_CONCURRENT_DOWNLOAD_TASKS,
+            max_workers=cfg.DOWNLOAD_THREADS,
+            thread_name_prefix="download_metadata_files",
+            initializer=self._downloader_workder_initializer,
+            watchdog_func=partial(
+                self._downloader_pool.downloading_watchdog,
+                ctx=self._download_watchdog_ctx,
+                max_idle_timeout=cfg.DOWNLOAD_INACTIVE_TIMEOUT,
+            ),
+        )
+
+        _condition = threading.Condition()
+        _metadata_processor = self._ota_metadata.download_metafiles(_condition)
+
+        try:
+            for _fut in _mapper.ensure_tasks(
+                partial(self._download_metadata_file, condition=_condition),
+                _metadata_processor,
+            ):
+                if not (_exc := _fut.exception()):
+                    continue
+
+                logger.warning(
+                    f"failed to download one metafile, keep retrying: {_exc!r}"
+                )
+                if isinstance(_exc, requests_exc.HTTPError) and isinstance(
+                    (_response := _exc.response), Response
+                ):
+                    if _response.status_code == HTTPStatus.NOT_FOUND:
+                        raise ota_errors.OTAImageInvalid(
+                            "failed to download metadata", module=__name__
+                        ) from _exc
+
+                    if _response.status_code in [
+                        HTTPStatus.FORBIDDEN,
+                        HTTPStatus.UNAUTHORIZED,
+                    ]:
+                        raise ota_errors.UpdateRequestCookieInvalid(
+                            module=__name__
+                        ) from _exc
```

Comment:
Previously when ota_metadata package does the metadata files downloading, we handle 404 and 403 errors there.
Now the ota_core takes the responsibility of handling 404 and 403 during metadata downloading, and we map the corresponding HTTP errors to OTA errors.